File: nodes/system-prompt/system-prompt.process.js
/**
 * Process function for the System Prompt node.
 * Outputs a message object, containing the prompt text and system role and text
 */
/**
 * Render a variable value for injection into prompt text.
 * Strings pass through untouched; everything else is JSON encoded so that
 * nested objects and arrays read as data instead of "[object Object]".
 */
const renderVariable = (value) => {
  if (value === null || value === undefined) return "";
  if (typeof value === "string") return value;
  try {
    return JSON.stringify(value, null, 2);
  } catch {
    // Circular structures (or anything else JSON can't encode) fall back to
    // the default coercion rather than failing the whole prompt.
    return String(value);
  }
};

export default async ({inputs, settings, config}) => {
  // Initialize variables array if not provided
  if(!inputs.variables) {
    inputs.variables = [];
  }

  // A single connection can deliver an array of key-value objects, so flatten
  // one level before looking keys up.
  const variables = (Array.isArray(inputs.variables) ? inputs.variables : [inputs.variables])
    .flatMap(entry => Array.isArray(entry) ? entry : [entry])
    .filter(entry => entry !== null && typeof entry === "object");

  // Get the base content from settings
  let baseContent = settings.content || "";

  // Handle chain input if provided
  let chainedContent = "";
  if (inputs.chain) {
    if (typeof inputs.chain === "string") {
      chainedContent = inputs.chain;
    } else if (typeof inputs.chain === "object" && inputs.chain.content) {
      // Handle message object format
      if (Array.isArray(inputs.chain.content)) {
        // Extract text content from array format
        chainedContent = inputs.chain.content
          .filter(item => item.type === "text")
          .map(item => item.text)
          .join("\n");
      } else if (typeof inputs.chain.content === "string") {
        chainedContent = inputs.chain.content;
      }
    }
  }

  // Combine chained content with base content
  let fullContent = chainedContent ? `${chainedContent}\n\n${baseContent}` : baseContent;

  // Create message object
  let message = {
    role: "system",
    content: [
      {
        type: "text",
        text: fullContent
      }
    ]
  };

  // Process variables
  message.content[0].text = message.content[0].text.replace(/\{\{(.*?)\}\}/g, (match, p1) => {
    // look for a variable with the key p1
    let variable = variables.find(variable => Object.keys(variable).find(key => key === p1));
    if(variable) {
      return renderVariable(variable[p1]);
    }
    return match;
  });
  
  // Return the message and string prompt
  return {
    message: message,
    prompt: message.content[0].text
  };
}; 

File: nodes/system-prompt/system-prompt.process.py
"""
System Prompt Node - Outputs a message object with the prompt text and system role.
"""

import json
import re
from typing import Any


def _render_variable(value: Any) -> str:
    """
    Render a variable value for injection into prompt text.
    Strings pass through untouched; everything else is JSON encoded so that
    nested objects and lists read as data instead of a Python repr.
    """
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    try:
        return json.dumps(value, indent=2)
    except (TypeError, ValueError):
        # Anything JSON can't encode falls back to the default coercion
        # rather than failing the whole prompt.
        return str(value)


def _flatten_variables(variables: Any) -> list:
    """
    A single connection can deliver a list of key-value objects, so flatten
    one level and keep only dicts.
    """
    if variables is None:
        return []
    if not isinstance(variables, list):
        variables = [variables]
    flat = []
    for entry in variables:
        if isinstance(entry, list):
            flat.extend(item for item in entry if isinstance(item, dict))
        elif isinstance(entry, dict):
            flat.append(entry)
    return flat


async def process(
    *,
    inputs: dict[str, Any],
    settings: dict[str, Any],
    config: dict[str, Any],
    node_config: dict[str, Any],
) -> dict[str, Any]:
    """
    Process function for the System Prompt node.
    Outputs a message object, containing the prompt text and system role.
    """
    # Initialize variables list if not provided
    variables = _flatten_variables(inputs.get("variables"))

    # Get the base content from settings
    base_content = settings.get("content", "")

    # Handle chain input if provided
    chained_content = ""
    chain = inputs.get("chain")
    if chain:
        if isinstance(chain, str):
            chained_content = chain
        elif isinstance(chain, dict) and chain.get("content"):
            # Handle message object format
            content = chain.get("content")
            if isinstance(content, list):
                # Extract text content from array format
                text_parts = [
                    item.get("text", "")
                    for item in content
                    if isinstance(item, dict) and item.get("type") == "text"
                ]
                chained_content = "\n".join(text_parts)
            elif isinstance(content, str):
                chained_content = content

    # Combine chained content with base content
    if chained_content:
        full_content = f"{chained_content}\n\n{base_content}"
    else:
        full_content = base_content

    # Create message object
    message = {
        "role": "system",
        "content": [
            {
                "type": "text",
                "text": full_content,
            }
        ],
    }

    # Process variables - replace {{key}} with variable value
    def replace_variable(match):
        key = match.group(1)
        # Look for a variable with the matching key
        for variable in variables:
            if isinstance(variable, dict) and key in variable:
                return _render_variable(variable[key])
        return match.group(0)  # Return original if no match

    message["content"][0]["text"] = re.sub(
        r"\{\{(.*?)\}\}",
        replace_variable,
        message["content"][0]["text"]
    )

    # Return the message and string prompt
    return {
        "message": message,
        "prompt": message["content"][0]["text"],
    }
